### DISCUSSION TITLE
Real World Examples?

### DISCUSSION THREAD

**Discussion by chabad360** [ergo-services/ergo#189] [Q&A]
The Ergo model is very interesting to me, and I would like to try building some stuff in it. But as someone who's completely unfamiliar with the methodology, it's really unclear how exactly certain things are supposed to work within this model. Stuff like database access (do I want to have a pool that sits in front of a meta-process? wait, I need a regular process to run the meta process, and how do I handle errors? Do I need to have a result type?), or API access (am I supposed to have a meta-process for the connection and then call into it?). Point is, I don't think these are going to be the only questions I run into, it took me a few hours to grok how stuff is supposed to get composed together, and I still have yet to figure out how to actually build stuff in this (and I'm pretty sure I'm not stupid).

Basically, this is a long-winded way of saying that having an example application(s) that has some degree of complexity would make it a lot easier for me (and I imagine a lot of other people) to learn the patterns here (which seem to be somewhat different from traditional Go).

**Comment by halturin** [comment]
Basically, you just need to understand the actor model itself. Its a bit different way of thinking :)

To work with DB, you do not need to use meta-process. It was created to work with things that require keeping read/write handler running (TCP socket, for example) while the actor handles messages sequentially. 

Think of your program design as a nano-service architecture in which actors are services that communicate to each other by message passing.

**Reply by chabad360** [reply]
I understand the model (I think), I've managed to divide my application into those small parts (For the most part). Where it breaks down for me is the implementation. 

I was expecting `Call` and `HandleCall` to pass the error without terminating the called actor. In the example of a DB query, say I want to return an error if there are no results found, that would terminate the DB process. Same goes for an HTTP api call, I'm not sure how I'm supposed to pass errors around? How many errors am I checking for (i.e. process error, returned error)? That's just one example, of something I'm not sure about. I wouldn't be surprised if the longer I spend with it, the more questions I'll have.

Maybe I'm still thinking too sequentially? In which case, having an actual application (think Todo MVC) I can look at would really go a long way for me.

**Reply by halturin** [reply]
what can i say, you didn't read the documentation

**Reply by chabad360** [reply]
> To work with DB, you do not need to use meta-process. It was created to work with things that require keeping read/write handler running (TCP socket, for example) while the actor handles messages sequentially.

Most databases I know of, you communicate with them over the network with a long lived TCP connection (often a pool of said connections). Not sure what exactly is different here?

**Reply by chabad360** [reply]
> what can i say, you didn't read the documentation

I did, numerous times, over and over. But after an additional reading, it's pretty clear I didn't do enough reading on the Actor Model. I'm doing that now. So... Fair enough.

However, none of that changes the fact, that instead of looking at some Go code and getting an intuitive understanding of how things want to work. I'm stuck either reading code in languages I'm not familiar with, or reading lots of articles.
If you have any suggested resources I should consult, feel free to drop them.

I think if I'm successful in wrapping my head fully around this, I will try to contribute some form of example application.

**Reply by halturin** [reply]
> I was expecting Call and HandleCall to pass the error without terminating the called actor


From the documentation

>For synchronous requests, the HandleCall callback method is invoked in an actor based on act.Actor. The result returned by this method is sent as the response to the request.

> To stop a process, simply return a non-nil error value from the message-handling callback. After termination, the Terminate callback method will be called, with the reason argument being the returned error value.


https://docs.ergo.services/actors/actor

**Reply by halturin** [reply]
> Most databases I know of, you communicate with them over the network with a long lived TCP connection (often a pool of said connections). Not sure what exactly is different here?

a huge difference. Working with DB you don't need to keep reading from the socket. You just send your request, receive the answer. That's it. 

The main idea with actors is to quickly handle received messages and return from the callback.

**Reply by halturin** [reply]
forgot to mention... there is a trick. if you really want to get the error as a result of Call request you may use `SendResponseError` of `gen.Process` interface. In this case, you have to `return nil, nil` in your `HandleCall` callback. 

But I wouldn't recommend to use this way since `error` for `Call` method implies the request has been delivered and handled by the actor. `any` there - for the result of such handling according to the business logic and could be literally anything - a value, or an error.

**Comment by Northward712** [comment]
As the original poster mentioned, I also want to understand actor-based solutions for databases like MongoDB or Redis. Should operations on the database be handled like other actors using Call or Send for synchronous or asynchronous data retrieval, or is it better to use global functions in the traditional Go way?

**Reply by Northward712** [reply]
> As the original poster mentioned, I also want to understand actor-based solutions for databases like MongoDB or Redis. Should operations on the database be handled like other actors using Call or Send for synchronous or asynchronous data retrieval, or is it better to use global functions in the traditional Go way?

I also use Erlang for game server development, and I use the [EMysql ](https://github.com/Eonblast/Emysql)database driver for MySQL operations. The general logic is to get a connection from a connection pool, then use spawn to asynchronously execute queries. The caller monitors the worker threads with Monitor and waits for the returned query data. I'm curious whether Ergo can also implement a similar Actor Pool approach to replace the aforementioned spawn --> receive model?

The execution code for EMysql is as follows:

```erlang
monitor_work(Connection, Timeout, {M, F, A}) when is_record(Connection, emysql_connection) ->
    %% spawn a new process to do work, then monitor that process until
    %% it either dies, returns data or times out.
    Parent = self(),
    Pid = spawn(
        fun() ->
            receive start ->
                Parent ! {self(), apply(M, F, A)}
            end
        end),
    Mref = erlang:monitor(process, Pid),
    Pid ! start,
    receive
        {'DOWN', Mref, process, Pid, {_, closed}} ->
            %-% io:format("monitor_work: ~p DOWN/closed -> renew~n", [Pid]),
            case emysql_conn:reset_connection(emysql_conn_mgr:pools(), Connection, keep) of
                NewConnection when is_record(NewConnection, emysql_connection) ->
                    %% re-loop, with new connection.
                    [_OldConn | RestArgs] = A,
                    NewA = [NewConnection | RestArgs],
                    monitor_work(NewConnection, Timeout, {M, F, NewA});
                {error, FailedReset} ->
                    exit({connection_down, {and_conn_reset_failed, FailedReset}})
            end;
        {'DOWN', Mref, process, Pid, Reason} ->
            %% if the process dies, reset the connection
            %% and re-throw the error on the current pid.
            %% catch if re-open fails and also signal it.
            %-% io:format("monitor_work: ~p DOWN ~p -> exit~n", [Pid, Reason]),
            case emysql_conn:reset_connection(emysql_conn_mgr:pools(), Connection, pass) of
                {error, FailedReset} ->
                    exit({Reason, {and_conn_reset_failed, FailedReset}});
                _ -> exit({Reason, {}})
            end;
        {Pid, Result} ->
            %% if the process returns data, unlock the
            %% connection and collect the normal 'DOWN'
            %% message send from the child process
            %-% io:format("monitor_work: ~p got result -> demonitor ~p, unlock connection ~p, return result~n", [Pid, Mref, Connection#emysql_connection.id]),
            erlang:demonitor(Mref, [flush]),
            emysql_conn_mgr:pass_connection(Connection),
            Result
    after Timeout ->
        %% if we timeout waiting for the process to return,
        %% then reset the connection and throw a timeout error
        %-% io:format("monitor_work: ~p TIMEOUT -> demonitor, reset connection, exit~n", [Pid]),
        erlang:demonitor(Mref),
        case emysql_conn:reset_connection(emysql_conn_mgr:pools(), Connection, pass) of
            {error, FailedReset} ->
                exit({mysql_timeout, Timeout, {and_conn_reset_failed, FailedReset}});
            _ -> exit({mysql_timeout, Timeout, {}})
        end
    end.
```

**Reply by halturin** [reply]
Basically, you dont have to do any special, just work with DB as you usually do.

```go
import (
    "ergo.services/ergo/act"
	"ergo.services/ergo/gen"

    "os"
    "database/sql"
    "github.com/go-sql-driver/mysql"
)

func factoryMyDB() gen.ProcessBehavior {
    return &myDB{}
}

type myDB struct {
	act.Actor
    db *sql.DB
}

type MyRequestSelect struct {
    F1 string
}

type MyF1 struct {
    A int
    B bool
}

func (m *myDB) Init(args ...any) error {
    cfg := mysql.Config{
        User:   os.Getenv("DBUSER"),
        Passwd: os.Getenv("DBPASS"),
        Net:    "tcp",
        Addr:   "127.0.0.1:3306",
        DBName: "exampleDB",
    }

    var err error
    db, err = sql.Open("mysql", cfg.FormatDSN())
    if err != nil {
        return err
    }

    if err := db.Ping(); err != nil {
        return err
    }

	m.Log().Info("started %s process on: %s", a.Name(), a.Node().Name())
	return nil
}


func (m *myDB) HandleCall(from gen.PID, ref gen.Ref, request any) (any, error) {
	switch r := request.(type) {
	case MyRequestSelect:
	        var f1s []MyF1
			m.Log().Info("received MyRequestSelect from %s: %#v", from, r)

	        rows, err := db.Query("SELECT a, b FROM tbl WHERE f1 = ?", r.F1)
	        if err != nil {
	            m.Log().Err("unable to SELECT data: %s", err)
	            return err, nil
 	       } 
  	      defer rows.Close()
  	      for rows.Next() {
 	           var f1 Album
   	         if err := rows.Scan(&f1.A, &f1.B); err != nil {
  	              return err, nil
 	           }
     	       f1s = append(f1s, f1)
   	     } 

  	      return f1s, nil
	}

	m.Log().Info("received unknown request: %#v", request)
	return nil, nil // do not reply on unknow requests
}

func (m *myDB) Terminate(reason error) {
	m.Log().Info("terminated with reason: %s", reason)
        m.db.Close()
}
```

**Reply by halturin** [reply]
example above doesnt handle db link disconnection (it just makes this process to terminate on error), but shows the idea.

**Reply by halturin** [reply]
>I'm curious whether Ergo can also implement a similar Actor Pool approach to replace the aforementioned spawn --> receive model?

sure, you can use `act.Pool` for this approach https://docs.ergo.services/actors/pool

**Reply by Northward712** [reply]
> > I'm curious whether Ergo can also implement a similar Actor Pool approach to replace the aforementioned spawn --> receive model?
> 
> sure, you can use `act.Pool` for this approach https://docs.ergo.services/actors/pool

Wow, that’s exactly how I also implemented it!

I previously tried using the code you mentioned to implement database operations, but I feel that in scenarios with many requests, a single actor can impact overall performance.

**Reply by Northward712** [reply]
> > > I'm curious whether Ergo can also implement a similar Actor Pool approach to replace the aforementioned spawn --> receive model?
> > 
> > 
> > sure, you can use  for this approach https://docs.ergo.services/actors/pool`act.Pool`
> 
> Wow, that’s exactly how I also implemented it!
> 
> I previously tried using the code you mentioned to implement database operations, but I feel that in scenarios with many requests, a single actor can impact overall performance.

However, during the process of using the pool, I encountered a problem: it cannot automatically set the RegisterName for each worker in the pool. I'm wondering if there will be an optimization in the future to assign a RegisterName to each worker in the pool.

**Reply by halturin** [reply]
yeah, `act.Pool` helps you to distribute the load.

**Reply by halturin** [reply]
>However, during the process of using the pool, I encountered a problem: it cannot automatically set the RegisterName for each worker

no need. register the name of your act.Pool process and use this name in your Call requests. see the docs.

**Reply by Northward712** [reply]
> yeah, `act.Pool` helps you to distribute the load.

Currently, the way I assign the RegisterName to workers is by getting their parent's RegisterName and then adding a random value. This approach isn't very intuitive.
